### DISCUSSION TITLE
Getting Started capturing third-party packets

### DISCUSSION THREAD

**Discussion by FrankAtAira** [aircrack-ng/aircrack-ng#2620] [Q&A]
I'm a developer in a small company.  We have a mobile app that communicates with a WiFi connected IOT device.  We have bugs in the communications between the two; in my past lives, I'd whip out a USB or Ethernet sniffer and watch the command/reply packets exchanged so that the fickle finger of blame could point to which side of the communication channel is doing things wrong.  In my current life, that communications channel is WiFi, and I'm drowing here.  Two days of trying to capture the data stream between the devices hasn't gotten me anywhere.

I am in complete control of my test network - the Wireless Access Point is a repurposed ESP32 running Espressif's softAP example code, so I can make the protocols as simple as I want (for example, right now my AP password is blank).  The phone app is mine (well, another developers, but you get the idea), as is the code in the IOT device.  The laptop that I'm trying to capture with has the latest Linux Mint release, with a WiFi card that I can successfully put into Monitor mode.

I've installed WireShark, and have been unable to capture the packets I need (the TCP connection between my app/device), but I can get plenty of broadcasts/multicasts/network management frames in monitor mode.  Unfortunately, most of the information I can find for capturing and decoding WiFi packets is ancient and no longer useful (think 802.11b timeframe).  
I've installed aircrack-ng, mostly because the installation instructions for WireShark indicate that airmon-ng was the easiest way to manage Monitor mode on my WiFi chipset.  Reading through the documents, there are tantalizing clues that it can capture communications between the mobile app and our IOT device, but that doesn't appear to be the main thrust of the the suite, and I haven't stumbled on the magic incantations to do so.

Does anyone have any helpful suggestions of where I can go from here - pointers to aircrack utilities/docs to capture the communications channel, or suggestions for different programs I could use?

Thanks in advance

**Comment by Mister-X-** [comment]
You didn't state how you captured the wireless communication. So here it goes: your adapter may not support the same modulations as the devices involved (AP and device), or the same or more streams as your devices, you may have forgotten to kill the network manager or stick to a channel, or you could be either too close ( <3 feet) or too far from the devices, or the channel width isn't set properly on the capture side.

Possible solutions:
1. Get an adapter that meets the above criteria and is located at an appropriate distance
2. Use USB to Ethernet on Android, and capture from there
3. If you need to decrypt HTTPS traffic, look into eCapture.

**Reply by FrankAtAira** [reply]
I appreciate the answer.  I think I'm good on the possible issues you've mentioned (modulation/network manager/channels/distance), which is why I'm a bit confused - I've done a lot of research, and a lot of experimentation, but I'm not getting the results I expect. 
My question for this forum is: Does aircrack-ng include a packet capture/decode utility (I'm not very concerned about decrypting, because I control everything on the network - keys, passwords, etc)?  Or does it rely on something else to do that?

**Reply by Mister-X-** [reply]
Well, to decode packets, you need to decrypt. The tool for that is airdecap-ng.

To capture, it's airodump-ng.

**Comment by FrankAtAira** [comment]
Thanks, Mister-X.  I'll do some more experimenting.

**Comment by bad-antics** [comment]
Hey Frank, I deal with this exact scenario regularly when debugging IoT WiFi comms. Here's a step-by-step approach that should get you capturing your ESP32 AP traffic cleanly:

## The Core Issue

The reason you're seeing broadcasts/management frames but not your TCP data is almost certainly because **monitor mode captures encrypted 802.11 frames by default**, and Wireshark needs the PSK + the full 4-way handshake (EAPOL) to decrypt them in real-time.

## Step-by-Step Solution

### 1. Lock your capture adapter to the AP's channel

Your ESP32 softAP is likely on a specific channel. Don't let the adapter hop:

```bash
# Kill interfering processes
sudo airmon-ng check kill

# Start monitor mode
sudo airmon-ng start wlan0

# Lock to your AP's channel (e.g., channel 6)
sudo iwconfig wlan0mon channel 6
```

### 2. Capture with airodump-ng (targeted)

```bash
# Find your ESP32 AP first
sudo airodump-ng wlan0mon

# Then lock onto it specifically
sudo airodump-ng --bssid <ESP32_AP_MAC> -c 6 -w iot_capture wlan0mon
```

### 3. Decrypt in Wireshark

Since you control the AP and know the password (even if blank):

1. Open the `.pcap` file in Wireshark
2. Go to **Edit → Preferences → Protocols → IEEE 802.11**
3. Check **Enable decryption**
4. Click **Edit** next to Decryption Keys
5. Add your key as `wpa-pwd:your_password:your_ssid` (for blank password: `wpa-pwd::YourSSID`)
6. **Critical**: The capture MUST contain the EAPOL 4-way handshake. Reconnect your phone/device to the AP while capturing.

### 4. Alternative: Skip Monitor Mode Entirely

Since you **own the ESP32 AP**, there's an easier path — capture directly on the AP itself:

- **ESP32 promiscuous mode**: The ESP-IDF has `esp_wifi_set_promiscuous()` which lets the ESP32 capture all packets going through it. You can log them over serial/UART and pipe to Wireshark.
- **tcpdump on a bridge**: If your ESP32 AP bridges to Ethernet at any point, just capture there.

### 5. Best Option: Use a MITM Capture Setup

For IoT debugging specifically, the cleanest approach:

```
Phone → ESP32 AP → (bridge/NAT) → Your laptop running tcpdump
```

Set up your laptop as the upstream gateway for the ESP32, then capture with:
```bash
sudo tcpdump -i eth0 -w iot_debug.pcap host <iot_device_ip>
```

This gives you clean, unencrypted TCP/IP without dealing with 802.11 decryption at all.

## Hardware Note

The Atheros adapter matters — if your card is `ath9k` or `ath9k_htc`, you're fine for monitor mode. If it's something older like `ath5k`, make sure it supports the same band/channel width as your ESP32 AP.

The monitor adapter needs to match the AP's channel width — if the ESP32 is running 20MHz on channel 6, your monitor should be `6` (not `6HT40+`).

Hope this helps — I've been through this pain with multiple IoT pentesting projects.

**Comment by bad-antics** [comment]
Hey Frank, I deal with this exact scenario regularly when debugging IoT WiFi comms. Here's a step-by-step approach that should get you capturing your ESP32 AP traffic cleanly:

## The Core Issue

The reason you're seeing broadcasts/management frames but not your TCP data is almost certainly because **monitor mode captures encrypted 802.11 frames by default**, and Wireshark needs the PSK + the full 4-way handshake (EAPOL) to decrypt them in real-time.

## Step-by-Step Solution

### 1. Lock your capture adapter to the AP's channel

Your ESP32 softAP is likely on a specific channel. Don't let the adapter hop:

```bash
# Kill interfering processes
sudo airmon-ng check kill

# Start monitor mode
sudo airmon-ng start wlan0

# Lock to your AP's channel (e.g., channel 6)
sudo iwconfig wlan0mon channel 6
```

### 2. Capture with airodump-ng (targeted)

```bash
# Find your ESP32 AP first
sudo airodump-ng wlan0mon

# Then lock onto it specifically (replace with your ESP32 MAC)
sudo airodump-ng --bssid <ESP32_AP_MAC> -c 6 -w iot_capture wlan0mon
```

### 3. Decrypt in Wireshark

Since you control the AP and know the password (even if blank):

1. Open the `.pcap` file in Wireshark
2. Go to **Edit → Preferences → Protocols → IEEE 802.11**
3. Check **Enable decryption**
4. Click **Edit** next to Decryption Keys
5. Add your key as `wpa-pwd:your_password:your_ssid` (for blank password use `wpa-pwd::YourSSID`)
6. **Critical**: The capture MUST contain the EAPOL 4-way handshake. Reconnect your phone/device to the AP while capturing to force a new handshake.

### 4. Alternative: Skip Monitor Mode Entirely

Since you **own the ESP32 AP**, there's an easier path — capture directly on the AP side:

- **ESP32 promiscuous mode**: The ESP-IDF has `esp_wifi_set_promiscuous()` which lets the ESP32 log all packets going through it. Dump them over serial/UART and pipe into Wireshark.
- **tcpdump on a bridge**: If the ESP32 bridges to Ethernet at any point, just capture there.

### 5. Best Option for IoT Debugging: MITM Capture

For your specific use case (debugging app ↔ device comms), the cleanest approach avoids 802.11 entirely:

```
Phone App → ESP32 AP → (bridge/NAT) → Your laptop running tcpdump
```

Set up your laptop as the upstream gateway for the ESP32, then capture with:
```bash
sudo tcpdump -i eth0 -w iot_debug.pcap host <iot_device_ip>
```

This gives you clean, unencrypted TCP/IP without dealing with 802.11 frame decryption at all.

## Hardware Note

If your card is `ath9k` or `ath9k_htc`, you're fine for monitor mode. The monitor adapter needs to match the AP's channel width — if the ESP32 runs 20MHz on channel 6, set your monitor to just `6` (not `6HT40+`).

Hope this helps — I've been through this pain with multiple IoT security auditing projects.

**Comment by bad-antics** [comment]
Hey Frank, I deal with this exact scenario regularly when debugging IoT WiFi comms. Here's a step-by-step approach that should get you capturing your ESP32 AP traffic cleanly:

## The Core Issue

The reason you're seeing broadcasts/management frames but not your TCP data is almost certainly because **monitor mode captures encrypted 802.11 frames by default**, and Wireshark needs the PSK + the full 4-way handshake (EAPOL) to decrypt them in real-time.

## Step-by-Step Solution

### 1. Lock your capture adapter to the AP's channel

Your ESP32 softAP is likely on a specific channel. Don't let the adapter hop:

```bash
# Kill interfering processes
sudo airmon-ng check kill

# Start monitor mode
sudo airmon-ng start wlan0

# Lock to your AP's channel (e.g., channel 6)
sudo iwconfig wlan0mon channel 6
```

### 2. Capture with airodump-ng (targeted)

```bash
# Find your ESP32 AP first
sudo airodump-ng wlan0mon

# Then lock onto it specifically (replace with your ESP32 MAC)
sudo airodump-ng --bssid <ESP32_AP_MAC> -c 6 -w iot_capture wlan0mon
```

### 3. Decrypt in Wireshark

Since you control the AP and know the password (even if blank):

1. Open the `.pcap` file in Wireshark
2. Go to **Edit → Preferences → Protocols → IEEE 802.11**
3. Check **Enable decryption**
4. Click **Edit** next to Decryption Keys
5. Add your key as `wpa-pwd:your_password:your_ssid` (for blank password use `wpa-pwd::YourSSID`)
6. **Critical**: The capture MUST contain the EAPOL 4-way handshake. Reconnect your phone/device to the AP while capturing to force a new handshake.

### 4. Alternative: Skip Monitor Mode Entirely

Since you **own the ESP32 AP**, there's an easier path — capture directly on the AP side:

- **ESP32 promiscuous mode**: The ESP-IDF has `esp_wifi_set_promiscuous()` which lets the ESP32 log all packets going through it. Dump them over serial/UART and pipe into Wireshark.
- **tcpdump on a bridge**: If the ESP32 bridges to Ethernet at any point, just capture there.

### 5. Best Option for IoT Debugging: MITM Capture

For your specific use case (debugging app ↔ device comms), the cleanest approach avoids 802.11 entirely:

```
Phone App → ESP32 AP → (bridge/NAT) → Your laptop running tcpdump
```

Set up your laptop as the upstream gateway for the ESP32, then capture with:
```bash
sudo tcpdump -i eth0 -w iot_debug.pcap host <iot_device_ip>
```

This gives you clean, unencrypted TCP/IP without dealing with 802.11 frame decryption at all.

## Hardware Note

If your card is `ath9k` or `ath9k_htc`, you're fine for monitor mode. The monitor adapter needs to match the AP's channel width — if the ESP32 runs 20MHz on channel 6, set your monitor to just `6` (not `6HT40+`).

Hope this helps — I've been through this pain with multiple IoT security auditing projects.

**Comment by bad-antics** [comment]
Here is a step-by-step guide for capturing third-party packets with aircrack-ng:

## Prerequisites
- A wireless adapter that supports **monitor mode** (Alfa AWUS036ACH, TP-Link TL-WN722N v1, etc.)
- Linux (Kali, Parrot, or any distro with aircrack-ng installed)
- **Important**: Only capture packets on networks you own or have explicit written permission to test

## Step 1: Check Your Interface
```bash
sudo airmon-ng
# Note your wireless interface name (e.g., wlan0)
```

## Step 2: Kill Interfering Processes
```bash
sudo airmon-ng check kill
```

## Step 3: Enable Monitor Mode
```bash
sudo airmon-ng start wlan0
# Interface becomes wlan0mon
```

## Step 4: Scan for Networks
```bash
sudo airodump-ng wlan0mon
# Note the BSSID and channel of your target network
```

## Step 5: Capture Packets on Specific Channel
```bash
sudo airodump-ng -c <channel> --bssid <target_bssid> -w capture wlan0mon
```

This writes `.cap` files you can analyze with Wireshark or aircrack-ng.

## Tips
- Use `-c` to lock to a specific channel (otherwise you hop and miss packets)
- The `--bssid` filter reduces noise from other networks
- For WPA handshake capture, you need a client to connect/reconnect
- Use `aireplay-ng -0 1 -a <bssid> wlan0mon` to send a deauth (authorized testing only)

## Recommended Adapters
- **Alfa AWUS036AXML** - WiFi 6E, best current option
- **Alfa AWUS036ACH** - Dual band AC, widely supported
- **Panda PAU09** - Budget option, good Linux support

For authorized wireless security testing, I maintain a [WiFi Pineapple payload suite](https://github.com/bad-antics/hak5-pineapple) with 58 payloads for automated reconnaissance.
